### PR TITLE
ENH:add autoreject channel thresholds to epoching

### PR DESCRIPTION
**Pull request by ktavabi**
@larsoner please review. And do you think you could test on funloc sample data once we agree on implementation?

### CODE REVIEW

File: mnefun/_mnefun.py [hunk 1]
```diff
@@ -77,6 +78,11 @@
 from mne.report import Report
 from mne.io.constants import FIFF
 
+try:
+    from autoreject import get_rejection_threshold
+except ImportError, e:
+    print(e)
```
Comment (larsoner) [inline mnefun/_mnefun.py]: Nest import to `autoreject` in whatever nested clause it will actually be used. Then no need for `try/except`

File: mnefun/_mnefun.py [hunk 2]
```diff
@@ -1380,7 +1390,21 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
                 warnings.warn('resulting new sampling frequency %s not equal '
                               'to previous values %s' % (new_sfreq, sfreqs))
             sfreqs.add(new_sfreq)
-        use_reject, use_flat = _restrict_reject_flat(p.reject, p.flat, raw)
+        _, use_flat = _restrict_reject_flat(p.reject, p.flat, raw)
```
Comment (larsoner) [inline mnefun/_mnefun.py]: This should go inside the `if p.autoreject_thresholds:`

File: mnefun/_mnefun.py [hunk 3]
```diff
@@ -1380,7 +1390,21 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
...
+            print('     Using autreject to compute rejection thresholds')
+            temp_epochs = Epochs(raw, events, event_id=None, tmin=p.tmin,
+                                 tmax=p.tmax, baseline=_get_baseline(p),
+                                 proj=True, reject=None, flat=use_flat,
+                                 preload=True)
+            new_dict = get_rejection_threshold(temp_epochs)
+            # only reject ecg events for MEG channels for now
+            #TODO test if ExG thresholds work during epoching
+            use_reject = {k: new_dict[k]
+                          for k in ['grad', 'mag'] if k in new_dict.keys()}
```
Comment (larsoner) [inline mnefun/_mnefun.py]: `['grad', 'mag', 'eeg']`

File: mnefun/_mnefun.py [hunk 4]
```diff
@@ -1824,6 +1848,7 @@ def do_preprocessing_combined(p, subjects, run_indices):
                   '        %s' % bad_file)
             if not op.isdir(bad_dir):
                 os.mkdir(bad_dir)
+            #TODO 86 autobad after autoreject merge
```
Comment (larsoner) [inline mnefun/_mnefun.py]: what does this mean?

File: mnefun/_mnefun.py [hunk 5]
```diff
@@ -1888,7 +1913,7 @@ def do_preprocessing_combined(p, subjects, run_indices):
 
         proj_nums = p.proj_nums
         eog_t_lims = [-0.25, 0.25]
-        ecg_t_lims = [-0.08, 0.08]
+        ecg_t_lims = [-0.2, 0.2]
```
Comment (larsoner) [inline mnefun/_mnefun.py]: why this change? It should probably be a parameter if it needs tweaking, as this is a backward-incompatible change

**Comment by kambysese** [comment]
I'll get to the rest of this tomorrow. I must have missed the email notification about your review. Thanks.

File: mnefun/_mnefun.py (hunk 4 above)
Comment (kambysese) [reply mnefun/_mnefun.py]: Thinking we could get rid of the automatic bad channel detection relying on arbitrary thresholds since autoreject module appears to also be able to "fix" bad channels.

Comment (larsoner) [reply mnefun/_mnefun.py]: We should leave all old methods present/functional for backward compatibility

File: mnefun/_mnefun.py (hunk 5 above)
Comment (kambysese) [reply mnefun/_mnefun.py]: I was just playing around with this and forgot to change it back. I was curious as to why you went with such a narrow window for QRS epoching? In the mne tutorials, the window is -500 - 500 ms and I've been using -200 - 200 ms for infants. Why is this backward-incompatible?

Comment (larsoner) [reply mnefun/_mnefun.py]: >  was curious as to why you went with such a narrow window for QRS epoching?

I don't remember but I doubt it makes a huge difference

> Why is this backward-incompatible?

It will change the computations and thus the results you get if you run `mnefun` before / after this PR (and not in a bug-fixing sort of way)

**Comment by kambysese** [comment]
Revised as per comments. Tested on funloc subject data...

image

yellow = epoching with `proj=True` + autoreject
red = epoching with `proj=False`  + autoreject
cyan = epoching with `proj=False` without autoreject

I am thinking we should set `proj=True` WDYT. I just noticed there is a conflict. Do you want to review and then I'll try to resolve before merge?

**Comment by larsoner** [comment]
> I am thinking we should set proj=True WDYT.

I think we should always use `proj='delayed'`, that way we can disable projs later if we want to.

File: mnefun/_mnefun.py [hunk 6]
```diff
@@ -1380,7 +1386,22 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
...
             sfreqs.add(new_sfreq)
-        use_reject, use_flat = _restrict_reject_flat(p.reject, p.flat, raw)
+        if p.autoreject_thresholds:
+            from autoreject import get_rejection_threshold
+            print('     Using autreject to compute rejection thresholds')
+            temp_epochs = Epochs(raw, events, event_id=None, tmin=p.tmin,
+                                 tmax=p.tmax, baseline=_get_baseline(p),
+                                 proj=False, reject=None, flat=None,
+                                 preload=True)
+            new_dict = get_rejection_threshold(temp_epochs, decim=2)
```
Comment (larsoner) [inline mnefun/_mnefun.py]: don't use `decim=2` here, use `decim[si]` in the `Epochs` constructor

File: mnefun/_mnefun.py [hunk 7]
```diff
@@ -1969,14 +1990,24 @@ def do_preprocessing_combined(p, subjects, run_indices):
                        **old_kwargs)
             raw.add_proj(projs)
             raw.apply_proj()
+            if p.autoreject_thresholds:
+                ecg_epochs = create_ecg_epochs(raw, tmin=ecg_t_lims[0],
+                                               tmax=ecg_t_lims[1],
+                                               ch_name=p.ecg_channel)
+                new_dict = get_rejection_threshold(ecg_epochs, decim=3)
```
Comment (larsoner) [inline mnefun/_mnefun.py]: same comment here

File: mnefun/_mnefun.py (hunk 1 above)
Comment (kambysese) [reply mnefun/_mnefun.py]:
```
...
    elif func == do_preprocessing_combined:
        if p.autoreject_thresholds:
            from autoreject import get_rejection_threshold
        outs[ii] = func(p, subjects, run_indices)
....
```

For the sake of not DRY'ing why does above snippet not work in `do_processing`. I can see the code go into the conditional and executing ` if p.autoreject_thresholds` as `True` but I still get 

> NameError: global name 'get_rejection_threshold' is not defined

down in `do_processing_combined`

Comment (larsoner) [reply mnefun/_mnefun.py]: It's a scope problem. When you import `get_rejection_threshold` in that function, the `func` (whatever it is, which must be `do_processing_combined`) does not have access to it. You should do the import in `do_processing_combined`.

Comment (larsoner) [reply mnefun/_mnefun.py]: (A good introspecting editor will give you a flake error in both functions, saying that `get_rejection_threshold` isn't used in the first function, and isn't known in the second)

Comment (kambysese) [reply mnefun/_mnefun.py]: Thanks. Pycharm has no native support for flake :disappointed:

**Comment by kambysese** [comment]
I am smelling a successful conflict resolution after two years of git --inon
Good to merge?

File: mnefun/_mnefun.py [hunk 8]
```diff
@@ -1406,7 +1414,23 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
                 warnings.warn('resulting new sampling frequency %s not equal '
                               'to previous values %s' % (new_sfreq, sfreqs))
             sfreqs.add(new_sfreq)
-        use_reject, use_flat = _restrict_reject_flat(p.reject, p.flat, raw)
+        if p.autoreject_thresholds:
+            if p.autoreject_thresholds:
```
Comment (larsoner) [inline mnefun/_mnefun.py]: you do not need this second conditional

File: mnefun/_mnefun.py [hunk 9]
```diff
@@ -1406,7 +1414,23 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
...
+                from autoreject import get_rejection_threshold
+            print('     Using autreject to compute rejection thresholds')
+            temp_epochs = Epochs(raw, events, event_id=None, tmin=p.tmin,
+                                 tmax=p.tmax, baseline=_get_baseline(p),
+                                 proj=True, reject=None, flat=None,
+                                 preload=True)
+            new_dict = get_rejection_threshold(temp_epochs, decim=decim[si])
```
Comment (larsoner) [inline mnefun/_mnefun.py]: do not pass `decim=decim[si]` here but rather in the `Epochs` constructor

**Comment by kambysese** [comment]
There is an algorithm to [find and repair bad channels](http://autoreject.github.io/generated/autoreject.Ransac.html#autoreject.Ransac) but I've not really played around with it. I am okay to merge global thresholding for now. It would be nice if there were some feedback from others on this stuff e.g., how things are working out for them.

**Comment by kambysese** [comment]
@larsoner Can we finish this too while we at it?

**Comment by larsoner** [comment]
Merge conflict

**Comment by kambysese** [comment]
Happy?

File: mnefun/_mnefun.py [hunk 10]
```diff
@@ -1411,7 +1419,22 @@ def save_epochs(p, subjects, in_names, in_numbers, analyses, out_names,
...
+            print('     Using autreject to compute rejection thresholds')
+            temp_epochs = Epochs(raw, events, event_id=None, tmin=p.tmin,
+                                 tmax=p.tmax, baseline=_get_baseline(p),
+                                 proj=True, reject=None, flat=None,
+                                 preload=True, decim=decim[si])
+            new_dict = get_rejection_threshold(temp_epochs)
+            use_reject, use_flat = _restrict_reject_flat(new_dict, p.flat, raw)
+            new_dict = get_rejection_threshold(temp_epochs)
+            use_reject = {k: new_dict[k]
+                          for k in ['grad', 'mag', 'eeg']
```
Comment (larsoner) [inline mnefun/_mnefun.py]: Should this also have `eog` and `ecg` in it? I think our `reject` can usually have at least `eog`, but I'm not sure if `autoreject` does anything with the non-M/EEG channels

Comment (kambysese) [reply mnefun/_mnefun.py]: So what happens if I use the entire rejection dictionary from autoreject e.g. `{'ecg': 2.93e-12, 'eog': 9.1e-05, 'eeg': 9.1e-05, 'grad': 7.97e-11, 'mag': 2.93e-12}` all epochs are rejected b/c ExG channels always exceed the autoreject criteria. In the case funloc subj_01 ECG063 causes all trials to be dropped. I assume this is because of how autoreject deals with ExG channels.

Comment (larsoner) [reply mnefun/_mnefun.py]: This sounds like a bug with `autoreject`. Can you replicate the problem with the MNE sample dataset?

Comment (kambysese) [reply mnefun/_mnefun.py]: I can run the question by Mainak, and maybe he'll look into it for us. But the other reason I opted for `['grad', 'mag', 'eeg']` is that IIRC in the past you have argued for using PCA to project out ExG artifacts rather than rejecting epochs based on their amplitudes.

Comment (larsoner) [reply mnefun/_mnefun.py]: We'll have to talk to the `autoreject` folks about how these thresholds are chosen. We do want to keep the artifacts of interest, but not e.g. gigantic channel discontinuity artifacts. It's possible that even using `eeg` here would try to omit most of the blink epochs, for example, depending on how `autoreject` works.

**Comment by kambysese** [comment]
WDYM by
"We do want to keep the artifacts of interest"
I posted a question to Mainak on gitter and CCed you.

**Comment by larsoner** [comment]
Blinks and heartbeat events

**Comment by kambysese** [comment]
Right. But what do you mean by keeping them?
I didn't get much of an explanation on gitter, but recent PR excluded EEG from threshold dictionary. What do you make of it?

**Comment by larsoner** [comment]
> But what do you mean by keeping them?

In at least some of the code of interest, we are creating epochs around artifacts (blink, heartbeat) so we want to keep them (not have them be rejected)

> I didn't get much of an explanation on gitter, but recent PR excluded EEG from threshold dictionary. What do you make of it?

What recent PR?

**Comment by kambysese** [comment]
> In at least some of the code of interest, we are creating epochs around artifacts (blink, heartbeat) so we want to keep them (not have them be rejected)

Ok. But that's not happening with `['grad', 'mag', 'eeg']` sub-selection from autoreject dictionary...I don't think.

[This PR](https://github.com/autoreject/autoreject/pull/81#issuecomment-334484599)...sorry markdown is a SOB on the phone.

File: mnefun/_mnefun.py [hunk 11]
```diff
@@ -2037,6 +2075,17 @@ def do_preprocessing_combined(p, subjects, run_indices):
                        **old_kwargs)
             raw.add_proj(projs)
             raw.apply_proj()
+            if p.autoreject_thresholds:
+                eog_epochs = create_eog_epochs(raw, tmin=eog_t_lims[0],
+                                               tmax=eog_t_lims[1],
+                                               ch_name=p.ecg_channel)
+                new_dict = get_rejection_threshold(eog_epochs)
```
Comment (larsoner) [inline mnefun/_mnefun.py]: Here and above we use `autoreject` to set the rejection params. We need to verify that it's not eliminating too many EOG epochs. Have you checked to see what proportion are removed? It should be fairly small for most datasets.

**Comment by kambysese** [comment]
@larsoner Can you review the following gist. In it, I am trying to see what proportions of EOG events are dropped with default rejection values from mne e.g., `dict(grad=4000e-13, mag=4e-12, eeg=500e-6, eog=250e-6)` vs. `{'eeg': 0.000174, 'eog': 0.000230, 'grad': 9.525e-11, 'mag': 3.01e-12}` from autoreject.

I am surprised that the number of epochs dropped by `compute_proj_eog` and `create_eog_epochs` using the same rejection criteria is not the same. Is that the expected behavior?
Secondly,  it looks like the criteria from autoreject will result in more epochs dropped when considering all channel types. Again there is a discrepancy between the number dropped epochs between `compute_proj_eog` and `create_eog_epochs`, with the latter being 2x as many with autoreject compared to arbitrary criteria. 

```
from mne.datasets.sample import data_path
from mne.io import read_raw_fif
from mne import read_events
from mne.preprocessing import compute_proj_eog, create_eog_epochs
from autoreject import get_rejection_threshold

data_path = data_path()
raw = read_raw_fif(data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif',
                   preload=True)
events = read_events(data_path + '/MEG/sample/sample_audvis_filt-0-40_raw-eve.fif')

pr, eog_events_a = compute_proj_eog(raw, n_grad=1,
                                    n_mag=1, n_eeg=1,
                                    tmin=-.25, tmax=.25,
                                    reject=dict(grad=4000e-13,
                                                mag=4e-12,
                                                eeg=500e-6,
                                                eog=250e-6))  # 12 eps drop

eog_epochs = create_eog_epochs(raw, reject=dict(grad=4000e-13,
                                                   mag=4e-12,
                                                   eeg=500e-6,
                                                   eog=250e-6))  # 16 eps drop  # noqa
eog_epochs.average().plot()

print('number of EOG events: %s' % len(eog_events_a))  # 46 events

# with autoreject
reject = get_rejection_threshold(eog_epochs)
print(reject)
pr, eog_events_b = compute_proj_eog(raw, n_grad=1,
                                    n_mag=1, n_eeg=1,
                                    tmin=-.25, tmax=.25,
                                    reject=reject)  # 15 eps drop

eog_epochs = create_eog_epochs(raw, reject=reject)  # 31 eps drop
eog_epochs.average().plot()
print('number of EOG events with autoreject: %s' % len(eog_events_b))  # 46 events
```

**Comment by larsoner** [comment]
> I am surprised that the number of epochs dropped by compute_proj_eog and create_eog_epochs using the same rejection criteria is not the same. Is that the expected behavior?

Looking at the API doc, [compute_proj_eog](http://mne-tools.github.io/dev/generated/mne.preprocessing.compute_proj_eog.html) filtetrs the data before processing and [create_eog_epochs](https://mne-tools.github.io/dev/generated/mne.preprocessing.create_eog_epochs.html) does not. Presumably if you filter `raw` before sending to the latter they will match.

**Comment by ktavabi** [comment]
> Looking at the API doc, compute_proj_eog filtetrs the data before processing and create_eog_epochs does not.

The default filtering args for `compute_proj_eog` are ` l_freq=1.0`, `h_freq=35.0`, and 1.0 & 10, respectively for `create_eog_epochs`. So I am not sure what you mean.

In hindsight tho it make very little sense to use autoreject to compute rejection criteria for artifact epochs, especially considering we want to maximize the number of epochs for input into SSP. I am reverting back to previous state. Here are results from funloc subj02 with red trace using autoreject dict and yellow using mnefun defaults.
autorej_testing

**Comment by larsoner** [comment]
Thanks @ktavabi
